protocol/binary/writer.go

```diff
@@ -201,32 +168,32 @@ func (bw *Writer) WriteValue(v wire.Value) error {
 	switch v.Type() {
 	case wire.TBool:
 		if v.GetBool() {
-			return bw.writeByte(1)
+			return bw.sw.writeByte(1)
 		}
-		return bw.writeByte(0)
+		return bw.sw.writeByte(0)
 
 	case wire.TI8:
-		return bw.writeByte(byte(v.GetI8()))
+		return bw.sw.writeByte(byte(v.GetI8()))
 
 	case wire.TDouble:
 		value := math.Float64bits(v.GetDouble())
-		return bw.writeInt64(int64(value))
+		return bw.sw.writeInt64(int64(value))
 
 	case wire.TI16:
-		return bw.writeInt16(v.GetI16())
+		return bw.sw.writeInt16(v.GetI16())
```

Comment:
We should avoid using private functionality of sw here. This can be implemented solely in terms of the exported streaming API.


```suggestion
		return bw.sw.WriteInt16(v.GetI16())
```